pkg/admission/apiexport/admission_test.go

```diff
@@ -0,0 +1,174 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package apiexport
+
+import (
+	"context"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/apiserver/pkg/admission"
+	"k8s.io/apiserver/pkg/authentication/user"
+
+	"github.com/kcp-dev/kcp/pkg/admission/helpers"
+	apisv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/apis/v1alpha1"
+)
+
+func createAttr(name string, obj runtime.Object, kind, resource string) admission.Attributes {
+	return admission.NewAttributesRecord(
+		helpers.ToUnstructuredOrDie(obj),
+		nil,
+		apisv1alpha1.Kind(kind).WithVersion("v1alpha1"),
+		"",
+		name,
+		apisv1alpha1.Resource(resource).WithVersion("v1alpha1"),
+		"",
+		admission.Create,
+		&metav1.CreateOptions{},
+		false,
+		&user.DefaultInfo{},
+	)
+}
+
+func updateAttr(name string, obj runtime.Object, kind, resource string) admission.Attributes {
+	return admission.NewAttributesRecord(
+		helpers.ToUnstructuredOrDie(obj),
+		nil,
+		apisv1alpha1.Kind(kind).WithVersion("v1alpha1"),
+		"",
+		name,
+		apisv1alpha1.Resource(resource).WithVersion("v1alpha1"),
+		"",
+		admission.Update,
+		&metav1.UpdateOptions{},
+		false,
+		&user.DefaultInfo{},
+	)
+}
+
+func TestAdmission(t *testing.T) {
+	cases := map[string]struct {
+		attr        admission.Attributes
+		update      bool
+		kind        string
+		resource    string
+		hasIdentity bool
+		isBuiltIn   func(apisv1alpha1.GroupResource) bool
+		want        error
+	}{
+		"NotAPIExportKind": {
+			kind:     "Something",
+			resource: "apiexports",
+			isBuiltIn: func(gr apisv1alpha1.GroupResource) bool {
+				return false
+			},
+		},
+		"NotAPIExportResource": {
+			kind:     "APIExport",
+			resource: "somethings",
+			isBuiltIn: func(gr apisv1alpha1.GroupResource) bool {
+				return false
+			},
+		},
+		"ForbiddenCreateNonInternalNoID": {
+			kind:     "APIExport",
+			resource: "apiexports",
+			isBuiltIn: func(gr apisv1alpha1.GroupResource) bool {
+				return false
+			},
+			want: field.Invalid(
+				field.NewPath("spec").
+					Child("permissionClaims").
+					Index(0).
+					Child("identityHash"),
+				"",
+				"identityHash is required for API types that are not built-in"),
+		},
+		"ForbiddenUpdateNonInternalNoID": {
+			update:   true,
+			kind:     "APIExport",
+			resource: "apiexports",
+			isBuiltIn: func(gr apisv1alpha1.GroupResource) bool {
+				return false
+			},
+			want: field.Invalid(
+				field.NewPath("spec").
+					Child("permissionClaims").
+					Index(0).
```

Comment:
Should probably add a test case that verifies a nonzero index?

Reply:
will do 👍🏻